json_codec.go

```diff
@@ -0,0 +1,846 @@
+//   // Decoding data uses codec created above, and an io.Reader,
+//   // definition not shown:
+//   datum, err := codec.Decode(r)
+//   if err != nil {
+//       return nil, err
+//   }
+//
+//   // Encoding data uses codec created above, an io.Writer,
+//   // definition not shown, and some data:
+//   err := codec.Encode(w, datum)
+//   if err != nil {
+//       return nil, err
+//   }
+//
+//   // Encoding data using bufio.Writer to buffer the writes
+//   // during data encoding:
+//
+//   func encodeWithBufferedWriter(c Codec, w io.Writer, datum interface{}) error {
+//	bw := bufio.NewWriter(w)
+//	err := c.Encode(bw, datum)
+//	if err != nil {
+//		return err
+//	}
+//	return bw.Flush()
+//   }
+//
+//   err := encodeWithBufferedWriter(codec, w, datum)
+//   if err != nil {
+//       return nil, err
+//   }
+func NewJSONCodec(someJSONSchema string, setters ...CodecSetter) (Codec, error) {
+	// unmarshal into schema blob
+	var schema interface{}
+	if err := json.Unmarshal([]byte(someJSONSchema), &schema); err != nil {
+		return nil, &ErrSchemaParse{"cannot unmarshal JSON", err}
+	}
+	// remarshal back into compressed json
+	compressedSchema, err := json.Marshal(schema)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal schema: %v", err)
+	}
+
+	// each codec gets a unified namespace of symbols to
+	// respective codecs
+	st := newJSONSymbolTable()
+
+	newCodec, err := st.buildCodec(nullNamespace, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, setter := range setters {
+		err = setter(newCodec)
+		if err != nil {
+			return nil, err
+		}
+	}
+	newCodec.schema = string(compressedSchema)
+	return newCodec, nil
+}
+
+func (st symtabJSON) buildCodec(enclosingNamespace string, schema interface{}) (*codec, error) {
+	switch schemaType := schema.(type) {
+	case string:
+		return st.buildString(enclosingNamespace, schemaType, schema)
+	case []interface{}:
+		return st.makeUnionCodec(enclosingNamespace, schema)
+	case map[string]interface{}:
+		return st.buildMap(enclosingNamespace, schema.(map[string]interface{}))
+	default:
+		return nil, newCodecBuildError("unknown", "schema type: %T", schema)
+	}
+}
+
+func (st symtabJSON) buildMap(enclosingNamespace string, schema map[string]interface{}) (*codec, error) {
+	t, ok := schema["type"]
+	if !ok {
+		return nil, newCodecBuildError("map", "ought have type: %v", schema)
+	}
+	switch t.(type) {
+	case string:
+		// EXAMPLE: "type":"int"
+		// EXAMPLE: "type":"enum"
+		return st.buildString(enclosingNamespace, t.(string), schema)
+	case map[string]interface{}, []interface{}:
+		// EXAMPLE: "type":{"type":fixed","name":"fixed_16","size":16}
+		// EXAMPLE: "type":["null","int"]
+		return st.buildCodec(enclosingNamespace, t)
+	default:
+		return nil, newCodecBuildError("map", "type ought to be either string, map[string]interface{}, or []interface{}; received: %T", t)
+	}
+}
+
+func (st symtabJSON) buildString(enclosingNamespace, typeName string, schema interface{}) (*codec, error) {
```

Comment:
It looks like a bunch of these methods for `symtabJSON` are exactly the same as `symtab`, as well as the struct definition. If it's because any of the symbol table methods are or need to be unique to `symtabJSON`, you could alternatively just use embedding and overload the methods you need to overload. 

I may be missing something, but wanna chat about this?